### ISSUE TITLE
Docs :: Add guidance on using loadClass with Typescript

### ISSUE THREAD

**Issue by sgpinkus** [Automattic/mongoose#12813]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I'm trying to port a JS project that uses `loadClass` to TS (it's doing my head in trying to get types working completely ..). The [documentation on TS](https://mongoosejs.com/docs/typescript.html) is good enough, except doesn't mention anything about `loadClass`. A basic example showing how to use it with Mongoose's typings, or alternatively, a statement recommending against its use, and briefly why would be helpful.

Aside: the fundamental class structure implied by loadClass seems to be at odds with Mongoose's actual class structure. Maybe this wasn't a problem in JS work, but it's definitely, at the least, a bit confusing in TS world. Maybe loadClass should be deprecated.

### Motivation

Help users trying to decide whether to try shoehorn loadClass into Mongoose typings or not.

**Comment by hasezoey** [comment]
as a note for those wanting information now instead of waiting for the documentation:

from what i can tell, `loadClass` currently does not affect the typescript type, so you would have to use the class like a interface and `&` it with the actual property interface

example:

```ts
class MyClass {
  myMethod() { return 42; }
  static myStatic() { return 42; }
  get myVirtual() { return 42; }
}

const schema = new mongoose.Schema({ property1: String });
schema.loadClass(MyClass);

type MySchema {
  property1: string;
}

type MyCombined = MySchema & MyClass;

const model = mongoose.model<MyCombined>('MyClass', schema);
```

**Comment by sgpinkus** [comment]
@hasezoey thanks.

Just to be clear, as far as I understand, there are "lean" caveats to above. Example: `doc.toObject().myMethod()` will compile and result in runtime error.

Main thing I wanted to do was type MyClass. The problem is "methods" and "statics" both have different sub classes in Mongoose / Mongoose type system. So the only way I've found to make it work is split the class in two: MyStatics, MyMethods, and call loadClass twice. All the methods in MyStatics have to be static for no reason which is a bit hacky, so maybe there should be `loadMethods()`, loadStatics()` to be consistent with how Mongoose actually works.

**Comment by hasezoey** [comment]
> Main thing I wanted to do was type MyClass. The problem is "methods" and "statics" both have different base classes in Mongoose / Mongoose type system. So the only way I've found to make it work is split the class in two: MyStatics, MyMethods, and call loadClass twice. All the methods in MyStatics have to be static for no reason which is a bit hacky, so maybe there should be loadMethods(), loadStatics()` to be consistent with how Mongoose actually works.

i am not sure what you mean with this, mongoose and typescript (and mongoose types) both correctly handle statics and instance methods just fine
in typescript you can access instance properties & methods via just `MyClass` and statics and static properties with `typeof MyClass`

just note that mongoose does not support defining properties in classes, only methods (both instance and static) and ES6 getter / setter

if you mean that the model type does not have the statics of your class, then you just need to do as shown in the example: `type MyCombinedModel = mongoose.Model<MyCombined> & typeof MyClass;`

**Comment by sgpinkus** [comment]
I just mean typing of `this` within methods MyClass. What I did was:

```
class MyStatics extends Model { ... }

class MyMethods extends Document { ... }

...

schema.loadClass(MyStatics}
schema.loadClass(MyMethods)
```

Even this is limited because `this` is still missing the `MySchema` typings.

> ... and statics and static properties with `typeof MyClass`

Ah yes, that's a useful tip to remember. Thanks.

**Comment by sgpinkus** [comment]
> i am not quite sure what you are asking; are you maybe asking how to set the correct this type for methods in the class?

Yes exactly. I'm aware you can set the type of this in the method signature, but it's a bit of a PITA when you have a class with 20 or so methods.

**Comment by hasezoey** [comment]
> but it's a bit of a PITA when you have a class with 20 or so methods.

that is true, but like i had said in my last comment, i dont think typescript supports redefining the `this` for the whole class, it (currently) needs to be done on a per-method basis (basically in every method)

**Comment by sgpinkus** [comment]
> i dont think typescript supports redefining the this

I doesn't, except via inheritance.

**Comment by dandean** [comment]
@hasezoey – `this` parameters do not appear to work on getters/setters. 

> 'get' and 'set' accessors cannot declare 'this' parameters.ts(2784)

**Comment by hasezoey** [comment]
@dandean thanks for pointing this out, i guess i didnt test `get` & `set` types specifically (or it may have worked in a previous typescript version, i dont recall)

related: https://github.com/microsoft/TypeScript/issues/52923

**Comment by Necro-Rohan** [comment]
Hi @hasezoey!
I noticed this documentation issue is still open and there’s a lot of helpful detail in the discussion thread. I’d be happy to contribute by summarizing the information into a new “**Using loadClass with TypeScript**” section for the docs.

Would a PR for this be helpful?

**Comment by hasezoey** [comment]
> Would a PR for this be helpful?

Sure go ahead, just know that some information may have become outdated as mongoose's types have advanced and became somewhat more complicated since then.

**Comment by Necro-Rohan** [comment]
Thanks for the clarification! I’ll proceed with drafting the PR and will keep the note about outdated typings in mind.

**Comment by Necro-Rohan** [comment]
Hi @hasezoey,
My PR (#15731) with the new documentation on loadClass + TypeScript is now open.
Whenever convenient, could you please review it?
I’m happy to make revisions based on your feedback.

Thank you!
